src/libsync/vfs/cfapi/cfapiwrapper.cpp

```diff
@@ -290,8 +301,134 @@ OCC::Optional<OCC::PinStateEnums::PinState> OCC::CfApiWrapper::PlaceHolderInfo::
     return cfPinStateToPinState(_data->PinState);
 }
 
-OCC::Result<void, QString> OCC::CfApiWrapper::registerSyncRoot(const QString &path, const QString &providerName, const QString &providerVersion)
+QString convertSidToStringSid(_In_ PSID sid)
```

Comment:
This feels like code copied and pasted from an MSN example. :-)

Could we make that safer and more ergonomic using higher level constructs. I spent quite some time in the rest of that file to avoid construct like the stringSid raw pointer, the calls to LocalFree and so on. There's value in doing this, it pays off in safety and debug dividends. :-)

This apply to the whole block here.

Reply:
> This feels like code copied and pasted from an MSN example. :-)
> 
> Could we make that safer and more ergonomic using higher level constructs. I spent quite some time in the rest of that file to avoid construct like the stringSid raw pointer, the calls to LocalFree, and so on. There's value in doing this, it pays off in safety and debug dividends. :-)
> 
> This apply to the whole block here.

@er-vin I tried to get rid of ::LocalFree(stringSid); for like 1 hour. Tried different approaches. QScopedPointer with custom deleted. QVector<wchar_t>, using "delete[] stringSid" instead of ::LocalFree". Nothing worked. It either leads to a heap corruption assert and a crash, either empty SID string or, just doesn't compile. Kindly suggest your vision of how to replace this code with a Qt-like alternative.
Having said that, I don't think using WinAPI ::Local free with raw pointer is a problem inside the Windows-specific implementation source file. If we want to get rid of WinAPI, I guess, we shouldn't be using this ::ConvertSidToStringSid function at all. Not to mentioned, we shouldn't even be using CfAPI as Microsoft base all their constructs on WinAPI :(

Reply:
Well, it's not *really* about getting rid of WinAPI, this would be pointless. It's more that their documentation and examples tend to push to use dangerous constructs because well... they don't have better tooling available when you use just that API. But we do and so we shall try to move away from the more dangerous constructs.

Reply:
@er-vin Alright. Well, I agree with you, but, I couldn't find any alternative to get rid of this construct. The problem here is, they expect a pointer to a pointer. So, they allocate memory and put it into that pointer internally. What I could do, is to try, and implement my own function `convertSidToStringSid` by shifting required bits to obtain the string representation of WindowsSid. But, isn't it even worse? With their function, at least, we can expect the conversion to always work from one version of Windows to another. In the case of re-inventing the wheel, we will lose this reliability.

P.S. I can also move the function into "Utility" if this will improve the situation.